controller/api/destination/external-workload/controller_util.go

```diff
@@ -0,0 +1,270 @@
+package externalworkload
+
+import (
+	"fmt"
+
+	ewv1alpha1 "github.com/linkerd/linkerd2/controller/gen/apis/externalworkload/v1alpha1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+)
+
+// getServicesToUpdateOnExternalWorkloadChange will look at an old and an
+// updated external workload resource and determine which services need to
+// be reconciled. The outcome is determined by what has changed in-between
+// resources (selections, spec, or both).
+func (ec *EndpointsController) getServicesToUpdateOnExternalWorkloadChange(o, c interface{}) ([]string, error) {
+	old, ok := o.(*ewv1alpha1.ExternalWorkload)
+	if !ok {
+		return nil, fmt.Errorf("couldn't get ExternalWorkload from object %#v", o)
+	}
+
+	updated, ok := c.(*ewv1alpha1.ExternalWorkload)
+	if !ok {
+		return nil, fmt.Errorf("couldn't get ExternalWorkload from object %#v", o)
+	}
+
+	labelsChanged := labelsChanged(old, updated)
+	specChanged := specChanged(old, updated)
+	if !labelsChanged && !specChanged {
+		ec.log.Debugf("skipping update; nothing has changed between old rv %s and new rv %s", old.ResourceVersion, updated.ResourceVersion)
+		return nil, nil
+	}
+
+	newSelection, err := ec.getExternalWorkloadSvcMembership(updated)
+	if err != nil {
+		return nil, err
+
+	}
+
+	oldSelection, err := ec.getExternalWorkloadSvcMembership(old)
+	if err != nil {
+		return nil, err
+	}
+
+	result := map[string]struct{}{}
+	// Determine the list of services we need to update based on the difference
+	// between our old and updated workload.
+	//
+	// Service selections (i.e. services that select a workload through a label
+	// selector) may reference an old workload, a new workload, or both,
+	// depending on the workload's labels.
+	if labelsChanged && specChanged {
```

Comment:
Having a hard time comparing this to https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/endpointslice/util/controller_utils.go#L184 and convincing myself that they're equivalent.  Can they be made more similar to make this easier?  A nice bonus of this is we get to use "k8s.io/apimachinery/pkg/util/sets" instead of hand rolled sets.

Reply:
will restructure